lib/Sema/MiscDiagnostics.cpp

```diff
@@ -3707,16 +3707,53 @@ static void diagnoseOptionalToAnyCoercion(TypeChecker &TC, const Expr *E,
               .highlight(subExpr->getSourceRange())
               .fixItInsertAfter(subExpr->getEndLoc(), " as Any");
         }
-      }
+      } else if (auto *literal = dyn_cast<InterpolatedStringLiteralExpr>(E)) {
+        // Warn about interpolated segments that contain optionals.
+        for (auto &segment : literal->getSegments()) {
+          // Allow explicit casts.
+          if (auto paren = dyn_cast<ParenExpr>(segment)) {
+            if (isa<ExplicitCastExpr>(paren->getSubExpr())) {
+              continue;
+            }
+          }
+
+          // Look through paren types.
+          auto segmentTy = segment->getType();
+          if (auto parenType = dyn_cast<ParenType>(segmentTy.getPointer())) {
+            segmentTy = parenType->getUnderlyingType();
+          }
 
```

Comment:
This isn't necessary. The call to getOptionalObjectType() strips the parens, and if you rework the diagnostics to print Types rather than StringRef I believe it strips parens as well.

Here is a quick tweak I just tried, to see what I mean: https://github.com/rudkx/swift/commit/6a43b76aeb5548d9f80bba98fdcbe4d5d58c79e7